File: scala/src/ch1/ex1-7.scala
object sqrt {
  val factor = 1000000.0

  def square (x: Double) : Double = x * x

  def average (x: Double, y: Double) : Double = (x + y) / 2.0

  def improve (guess: Double, x: Double) : Double = average (guess, x / guess)

  def is_good_enough (guess: Double, x: Double) : Boolean = Math.abs (square(guess) - x) < (guess / factor)

  def sqrt_iter (guess: Double, x: Double) : Double = {
    if (is_good_enough(guess, x)) guess else sqrt_iter(improve(guess, x), x)
  }

  def apply (x: Double) : Double = sqrt_iter(1.0, x)
}

object ex1_7 extends App {
  val delta = 0.000001
  assert (Math.abs (sqrt.square(sqrt.apply(0.02)) - 0.02) < delta)
  assert (Math.abs (sqrt.square(sqrt.apply(0.0002)) - 0.0002) < delta)
  println("done testing ex1_7")
}

File: javascript/ch1/ex1-7.js
load('../functions.js');

sq = function (x) {
    return x * x;
};

average = function (x, y) {
    return (x + y) / 2.0;
};

improve = function (guess, x) {
    return average(guess, x / guess);
};

is_good_enough = function (guess, x) {
    return Math.abs(sq(guess) - x) < (guess / 100000);
};

sqrt_iter = function (guess, x) {
    return is_good_enough(guess, x) ? guess : sqrt_iter(improve(guess, x), x);
};

sqrt = function (x) {
    return sqrt_iter(1.0, x);
};

var delta = 0.0001;
check_expect_float(sqrt(0.02), Math.sqrt(0.02), delta);
check_expect_float(sqrt(0.0002), Math.sqrt(0.0002), delta);
